doc/src/explanation/active-deprecations.md

```diff
@@ -0,0 +1,1231 @@
+with the previous Smith Normal Form code should switch to using
+`DomainMatrix` as shown in issue
+[#21402](https://github.com/sympy/sympy/pull/21402). A better API for the
+Smith normal form will be added later.
+
+(deprecated-non-expr-in-matrix)=
+### Non-`Expr` objects in a Matrix
+
+In SymPy 1.8 and earlier versions it was possible to put non-{class}`~.Expr`
+elements in a [`Matrix`](sympy.matrices.dense.Matrix) and the matrix elements could be any arbitrary
+Python object:
+
+```python
+>>> M = Matrix([[(1, 2), {}]]) # doctest: +SKIP
+>>> M # doctest: +SKIP
+```
+
+This is not useful and does not really work e.g.:
+
+```python
+>>> M + M # doctest: +SKIP
+Traceback (most recent call last):
+...
+TypeError: unsupported operand type(s) for +: 'Dict' and 'Dict'
+```
+
+The main reason for making this possible was that there were a number of
+`Matrix` subclasses in the SymPy codebase that wanted to work with objects
+from the polys module e.g.
+
+1. `RawMatrix` (see [above](deprecated-rawmatrix)) was used in `solve_lin_sys` which was part of `heurisch` and was also used by `smith_normal_form`. The `NewMatrix` class used domain elements as the elements of the Matrix rather than `Expr`.
+
+2. `NewMatrix` was used in the `holonomic` module and also used domain
+elements as matrix elements
+
+3. `PolyMatrix` used a mix of `Poly` and `Expr` as the matrix elements and was
+used by `risch`.
+
+All of these matrix subclasses were broken in different ways and the
+introduction of {class}`~.DomainMatrix`
+([#20780](https://github.com/sympy/sympy/issues/20780),
+[#20759](https://github.com/sympy/sympy/issues/20759),
+[#20621](https://github.com/sympy/sympy/issues/20621),
+[#19882](https://github.com/sympy/sympy/issues/19882),
+[#18844](https://github.com/sympy/sympy/issues/18844)) provides a better
+solution for all cases. Previous PRs have removed the dependence of these
+other use cases on Matrix
+([#21441](https://github.com/sympy/sympy/issues/21441),
+[#21427](https://github.com/sympy/sympy/issues/21427),
+[#21402](https://github.com/sympy/sympy/issues/21402)) and now
+[#21496](https://github.com/sympy/sympy/issues/21496) has deprecated having
+non-`Expr` in a `Matrix`.
+
+This change makes it possible to improve the internals of the Matrix class but
+it potentially impacts on some downstream use cases that might be similar to
+the uses of Matrix with non-Expr elements that were in the SymPy codebase. A
+potential replacement for code that used Matrix with non-Expr elements is
+DomainMatrix if the elements are something like domain elements and a domain
+object can be provided for them. Alternatively if the goal is just printing
+support then perhaps TableForm can be used.
+
+It isn't clear what to advise as a replacement here without knowing more about
+the usecase. Feel free to comment below indicating what you want to do with
+non-Expr Matrix elements and we can discuss how to achieve that.
```

Comment:
What does "feel free to comment below" mean if this is now in the docs rather than a GH issue?